### PR TITLE
refactor: karma contract extracted to a separate one

### PR DESCRIPTION
**Pull request by kupermind**
- Karma contract extracted to a separate one.

### CODE REVIEW

File: contracts/Karma.sol [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    function changeImplementation(address newImplementation) external {
+        // Check for the ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        if (newImplementation == address(0)) {
+            revert();
```
Comment (77ph) [inline contracts/Karma.sol:45]: TODO: error message

File: contracts/Karma.sol [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        // Store the karma implementation address
+        assembly {
+            sstore(KARMA_PROXY, newImplementation)
+        }
```
Comment (77ph) [inline contracts/Karma.sol:51]: TODO: event

File: contracts/Karma.sol [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    function setMechMarketplaceStatuses(address[] memory mechMarketplaces, bool[] memory statuses) external {
+        // Check for the ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        if (mechMarketplaces.length != statuses.length) {
+            revert();
```
Comment (77ph) [inline contracts/Karma.sol:78]: TODO: error message

File: contracts/Karma.sol [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        for (uint256 i = 0; i < mechMarketplaces.length; ++i) {
+            if (mechMarketplaces[i] == address(0)) {
+                revert();
+            }
+
+            mapMechMarketplaces[mechMarketplaces[i]] = statuses[i];
+        }
```
Comment (77ph) [inline contracts/Karma.sol:87]: TODO: event

File: contracts/Karma.sol [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    function changeMechKarma(address mech, int256 karma) external {
+        if (!mapMechMarketplaces[msg.sender]) {
+            revert();
```
Comment (77ph) [inline contracts/Karma.sol:92]: TODO: error message

File: contracts/Karma.sol [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        mapMechKarma[mech] += karma;
+    }
```
Comment (77ph) [inline contracts/Karma.sol:96]: TODO: event

File: contracts/Karma.sol [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    function increaseRequesterMechKarma(address requester, address mech) external {
+        if (!mapMechMarketplaces[msg.sender]) {
+            revert();
```
Comment (77ph) [inline contracts/Karma.sol:100]: TODO: error message

File: contracts/Karma.sol [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        mapRequesterMechKarma[requester][mech]++;
```
Comment (77ph) [inline contracts/Karma.sol:103]: TODO: event

File: contracts/KarmaProxy.sol [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            returndatacopy(0, 0, returndatasize())
+            if eq(success, 0) {
+                revert(0, returndatasize())
+            }
+            return(0, returndatasize())
+        }
+    }
+
+    /// @dev Gets the implementation address.
+    function getImplementation() external view returns (address implementation) {
```
Comment (77ph) [inline contracts/KarmaProxy.sol:54]: move to implementation?

**Review by 77ph — APPROVED** [review]

File: contracts/Karma.sol [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    function changeMechKarma(address mech, int256 karma) external {
+        if (!mapMechMarketplaces[msg.sender]) {
+            revert();
+        }
+
+        mapMechKarma[mech] += karma;
+    }
+
+    function increaseRequesterMechKarma(address requester, address mech) external {
```
Comment (DavidMinarsch) [inline contracts/Karma.sol:98]: can we also make this "changeRequesterMechKarma" and provide the increment/decrement value in the method

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**